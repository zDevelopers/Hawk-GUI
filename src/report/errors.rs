use std::result::Result;

use uuid::Uuid;

pub type ReportResult<T> = Result<T, InvalidReportError>;

#[derive(Debug, Fail, AsRefStr)]
pub enum InvalidReportError {
    #[fail(
        display = "No player with this UUID can be found in the players list: {}",
        uuid
    )]
    MissingPlayerReference { uuid: Uuid },

    #[fail(display = "An unknown error happened")]
    Unknown,
}
