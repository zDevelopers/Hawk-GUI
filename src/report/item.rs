#[inline(always)]
fn default_one_u8() -> u8 {
    1
}

#[inline(always)]
fn default_one_u32() -> u32 {
    1
}

/// A Minecraft item.
#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq, Hash)]
pub struct Item {
    /// The Minecraft item identifier.
    pub id: String,

    /// How many of them?
    #[serde(rename = "Count", default = "default_one_u8")]
    pub count: u8,

    /// Item metadata (if any)
    pub tag: Option<ItemTag>
}

/// Non-exhaustive item tags (the ones we need to know about) as of
/// https://minecraft.gamepedia.com/Player.dat_format#Item_structure .
#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq, Hash)]
#[serde(rename_all = "PascalCase")]
pub struct ItemTag {
    pub unbreakable: Option<u8>,  // Technically a boolean but stored as 0/1
    pub enchantments: Option<Vec<EnchantmentTag>>,
    pub stored_enchantments: Option<Vec<EnchantmentTag>>,
    pub custom_potion_effects: Option<Vec<PotionEffectTag>>,
    pub potion: Option<String>,
    #[serde(rename = "display")]
    pub display: Option<DisplayTag>
}

/// An enchantment applied to an item.
#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq, Hash)]
pub struct EnchantmentTag {
    pub id: String,
    #[serde(default = "default_one_u32")]
    pub lvl: u32  // As silly as this could sound, the maximal enchantment level (for 1.14+) is 2,147,483,647
}

/// A potion effect (ignoring fields we don't care about, i.e. display-related ones).
#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq, Hash)]
#[serde(rename_all = "PascalCase")]
pub struct PotionEffectTag {
    pub id: u8,
    pub amplifier: i8,
    #[serde(default = "default_one_u32")]
    pub duration: u32
}

/// The display tag of an item
#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq, Hash)]
#[serde(rename_all = "PascalCase")]
pub struct DisplayTag {
    pub name: Option<String>,
    pub lore: Option<Vec<String>>
}
