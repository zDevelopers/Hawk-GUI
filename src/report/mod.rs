use std::fs::{create_dir_all, File};
use std::io::{Read, Write};
use std::iter;
use std::path::Path;

use failure::{Error, ResultExt};
use rand::{Rng, thread_rng};
use rand::distributions::Alphanumeric;
use serde_json;

pub mod raw;
pub mod settings;
pub mod player;
pub mod team;
pub mod damage;
pub mod heal;
pub mod event;
pub mod aggregates;
pub mod report;
pub mod errors;

static PROCESSED_REPORT_FILE_NAME: &'static str = "report.json";
static RAW_REPORT_FILE_NAME: &'static str = "raw-report.json";

///
/// Reads a raw JSON report to the raw Rust structure.
///
pub fn read_raw_report<R: Read>(reader: R) -> Result<raw::Report, Error> {
    Ok(serde_json::from_reader(reader).context("Unable to parse raw report JSON")?)
}

///
/// Reads a processed JSON report to the processed Rust structure.
///
pub fn read_processed_report<R: Read>(reader: R) -> Result<report::Report, Error> {
    Ok(serde_json::from_reader(reader).context("Unable to parse processed report JSON")?)
}

///
/// Reads a processed report from its slug into the processed Rust structure.
///
pub fn read_processed_report_from_slug(slug: String) -> Result<report::Report, Error> {
    let raw_path = &path_for_slug(slug);
    let path = Path::new(raw_path);
    let file = File::open(path.join(PROCESSED_REPORT_FILE_NAME)).with_context(|_| format!("Unable to open report at {}", raw_path))?;
    Ok(read_processed_report(file)?)
}

///
/// From a raw report structure, process and save both raw & processed JSON
/// into the file system. Returns the slug to use to access the report, or
/// the error.
///
pub fn save_report(raw_report: raw::Report) -> Result<String, Error> {
    let mut rng = thread_rng();

    let report_slug = loop {
        let report_slug: String = iter::repeat(())
            .map(|()| rng.sample(Alphanumeric))
            .take(8).collect();

        if Path::new(&path_for_slug(report_slug.clone())).exists() {
            continue;
        }

        break report_slug
    };

    let root = path_for_slug(report_slug.clone());
    let root_path = Path::new(&root);
    let raw_path = root_path.join(RAW_REPORT_FILE_NAME);
    let processed_path = root_path.join(PROCESSED_REPORT_FILE_NAME);

    create_dir_all(root).expect("Unable to create directory to store the report");

    let mut raw_file = File::create(raw_path)?;
    raw_file.write_all(serde_json::to_string(&raw_report)?.as_bytes())?;

    let report = report::Report::from_raw(raw_report)?;

    let mut processed_file = File::create(processed_path)?;
    processed_file.write_all(serde_json::to_string(&report)?.as_bytes())?;

    Ok(report_slug)
}

fn path_for_slug<'a>(slug: String) -> String {
    format!(
        "{f}/reports/{prefix}/{slug}",
        f = crate::USERS_CONTENT_FOLDER,
        prefix = &slug[..2],
        slug = &slug
    )
}
