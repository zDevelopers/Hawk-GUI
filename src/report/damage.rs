use std::collections::HashMap;
use std::rc::Rc;
use std::time::Duration;

use chrono::{DateTime, FixedOffset};
use uuid::Uuid;

use crate::report::errors::{InvalidReportError, ReportResult};
use crate::report::player::{Player, SimplePlayer};
use crate::report::raw::Damage as RawDamage;
use crate::report::report::since;

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Damage {
    pub date: DateTime<FixedOffset>,
    pub since_beginning: Duration,
    pub cause: DamageCause,
    pub weapon: Weapon,
    pub weapon_enchantments: HashMap<String, u32>, // As silly as this could sound, the maximal enchantment level (for 1.14+) is 2,147,483,647â€Œ
    pub damager: Option<SimplePlayer>,
    pub damagee: SimplePlayer,
    pub damage: u16,
    pub lethal: bool
}

impl Damage {
    pub fn from_raw(raw_damage: &RawDamage, players: &HashMap<Uuid, Rc<Player>>, begin: &DateTime<FixedOffset>) -> ReportResult<Self> {
        match players.get(&raw_damage.damagee) {
            Some(damagee) => Ok(Self {
                date: raw_damage.date.clone(),
                since_beginning: since(&raw_damage.date, begin),
                cause: raw_damage.cause.clone(),
                weapon: raw_damage.weapon.clone().unwrap_or(Weapon::Fists).clone(),
                weapon_enchantments: match &raw_damage.weapon_enchantments {
                    Some(enchantments) => enchantments.clone(),
                    None => HashMap::new()
                },
                damager: match raw_damage.damager {
                    Some(damager) => match players.get(&damager) {
                        Some(damager) => Some((*damager).as_ref().into()),
                        None => Err(InvalidReportError::MissingPlayerReference { uuid: raw_damage.damagee })?
                    },
                    None => None
                },
                damagee: (*damagee).as_ref().into(),
                damage: raw_damage.damage,
                lethal: raw_damage.lethal
            }),
            None => Err(InvalidReportError::MissingPlayerReference { uuid: raw_damage.damagee })
        }
    }

    pub fn from_raw_vec(raw_damages: &Vec<RawDamage>, players: &HashMap<Uuid, Rc<Player>>, begin: &DateTime<FixedOffset>) -> ReportResult<Vec<Self>> {
        let mut damages = Vec::new();

        for damage in raw_damages {
            damages.push(Self::from_raw(damage, players, begin)?);
        }

        damages.sort_by(|a, b| a.date.cmp(&b.date));

        Ok(damages)
    }
}

#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq, Hash)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum DamageCause {
    Player,
    Zombie,
    Skeleton,
    Pigman,
    Witch,
    Spider,
    CaveSpider,
    Creeper,
    Enderman,
    Slime,
    Ghast,
    MagmaCube,
    Blaze,
    Wolf,
    AngryWolf,
    Silverfish,
    IronGolem,
    ZombieVillager,
    EnderDragon,
    Wither,
    WitherSkeleton,

    Fire,
    Lava,
    Thunderbolt,
    Cactus,
    TNT,
    Fall,
    Suffocation,
    Drowning,
    Starvation,

    Command,
    Unknown
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum Weapon {
    Fists,

    SwordWood,
    SwordStone,
    SwordIron,
    SwordGold,
    SwordDiamond,

    AxeWood,
    AxeStone,
    AxeIron,
    AxeGold,
    AxeDiamond,

    Bow,

    Magic,
    Thorns,

    Unknown
}

impl DamageCause {
    pub fn is_creature(&self) -> bool {
        match self {
            DamageCause::Player => true,
            DamageCause::Zombie => true,
            DamageCause::Skeleton => true,
            DamageCause::Pigman => true,
            DamageCause::Witch => true,
            DamageCause::Spider => true,
            DamageCause::CaveSpider => true,
            DamageCause::Creeper => true,
            DamageCause::Enderman => true,
            DamageCause::Slime => true,
            DamageCause::Ghast => true,
            DamageCause::MagmaCube => true,
            DamageCause::Blaze => true,
            DamageCause::Wolf => true,
            DamageCause::AngryWolf => true,
            DamageCause::Silverfish => true,
            DamageCause::IronGolem => true,
            DamageCause::ZombieVillager => true,
            DamageCause::EnderDragon => true,
            DamageCause::Wither => true,
            DamageCause::WitherSkeleton => true,

            DamageCause::Fire => false,
            DamageCause::Lava => false,
            DamageCause::Thunderbolt => false,
            DamageCause::Cactus => false,
            DamageCause::TNT => false,
            DamageCause::Fall => false,
            DamageCause::Suffocation => false,
            DamageCause::Drowning => false,
            DamageCause::Starvation => false,
            DamageCause::Command => false,
            DamageCause::Unknown => false,
        }
    }
}
