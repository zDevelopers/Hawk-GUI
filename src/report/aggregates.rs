use std::collections::{BTreeMap, HashMap};
use std::rc::Rc;
use std::time::Duration;

use chrono::{DateTime, FixedOffset};
use uuid::Uuid;

use crate::report::damage::{Damage, DamageCause};
use crate::report::heal::Heal;
use crate::report::player::{Player, PlayerStatistics, DisplayedPlayerStatistics, SimplePlayer};
use crate::report::report::since;
use crate::report::settings::SettingsPlayers;

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Aggregate {
    pub global_statistics: DisplayedPlayerStatistics,
    pub players_damages: BTreeMap<Uuid, PlayerAlterationsAggregate>,
    pub environmental_damages: EnvironmentalDamagesAggregate,
}

impl Aggregate {
    pub fn from_raw(
        players: &HashMap<Uuid, Rc<Player>>,
        damages: &Vec<Damage>,
        heals: &Vec<Heal>,
        begin: &DateTime<FixedOffset>,
        settings: &SettingsPlayers,
    ) -> Self {
        Aggregate {
            global_statistics: Self::aggregate_global_statistics(
                &players
                    .iter()
                    .filter_map(|(_uuid, player)| (player.as_ref()).statistics.clone())
                    .collect(),
                settings
            ),
            players_damages: Self::aggregate_alterations(players, damages, heals, begin),
            environmental_damages: Self::aggregate_environmental_damages(damages),
        }
    }

    fn aggregate_global_statistics(statistics: &Vec<PlayerStatistics>, settings: &SettingsPlayers) -> DisplayedPlayerStatistics {
        DisplayedPlayerStatistics::calculate_displayed_statistics(&PlayerStatistics {
            generic: Some(Self::aggregate_single_statistic_group(
                &statistics
                    .iter()
                    .filter_map(|statistic| statistic.generic.clone())
                    .collect(),
            )),
            used: Some(Self::aggregate_single_statistic_group(
                &statistics
                    .iter()
                    .filter_map(|statistic| statistic.used.clone())
                    .collect(),
            )),
            mined: Some(Self::aggregate_single_statistic_group(
                &statistics
                    .iter()
                    .filter_map(|statistic| statistic.mined.clone())
                    .collect(),
            )),
            picked_up: Some(Self::aggregate_single_statistic_group(
                &statistics
                    .iter()
                    .filter_map(|statistic| statistic.picked_up.clone())
                    .collect(),
            )),
        }, settings)
    }

    fn aggregate_single_statistic_group(
        statistics: &Vec<BTreeMap<String, u32>>,
    ) -> BTreeMap<String, u32> {
        let mut aggregated = BTreeMap::new();

        statistics
            .iter()
            .cloned()
            .flatten()
            .for_each(|(stat, val)| {
                aggregated.insert(
                    stat.clone(),
                    match aggregated.get(&stat) {
                        Some(prev_value) => prev_value + val,
                        None => val,
                    },
                );
            });

        aggregated
            .into_iter()
            .filter(|(_stat, val)| val != &0u32)
            .collect()
    }

    fn aggregate_alterations(
        players: &HashMap<Uuid, Rc<Player>>,
        damages: &Vec<Damage>,
        heals: &Vec<Heal>,
        begin: &DateTime<FixedOffset>,
    ) -> BTreeMap<Uuid, PlayerAlterationsAggregate> {
        let ranks: BTreeMap<Uuid, u8> = {
            let mut lethal_damages: Vec<&Damage> =
                damages.iter().filter(|damage| damage.lethal).collect();
            lethal_damages.sort_by(|a, b| a.date.cmp(&b.date).reverse());

            let ranked_deads: Vec<Uuid> = lethal_damages
                .iter()
                .map(|damage| damage.damagee.uuid)
                .collect();

            players
                .iter()
                .map(|(uuid, _player)| uuid)
                .filter(|uuid| !ranked_deads.contains(uuid))
                .chain(ranked_deads.iter())
                .enumerate()
                .map(|(index, uuid)| (uuid.clone(), index as u8 + 1))
                .collect()
        };

        players
            .iter()
            .map(|(uuid, _player)| {
                (
                    uuid.clone(),
                    Self::aggregate_player_alterations(
                        uuid,
                        damages,
                        heals,
                        *ranks.get(uuid).unwrap_or(&0u8),
                        begin,
                    ),
                )
            })
            .collect()
    }

    fn aggregate_player_alterations(
        player: &Uuid,
        damages: &Vec<Damage>,
        heals: &Vec<Heal>,
        rank: u8,
        begin: &DateTime<FixedOffset>,
    ) -> PlayerAlterationsAggregate {
        let mut damages_taken: Vec<Damage> = damages
            .iter()
            .cloned()
            .filter(|damage| &damage.damagee.uuid == player)
            .collect();
        let mut damages_caused: Vec<Damage> = damages
            .iter()
            .cloned()
            .filter(|damage| match &damage.cause {
                DamageCause::Player(cause) => &cause.player.uuid == player,
                _ => false
            })
            .collect();
        let mut heals: Vec<Heal> = heals
            .iter()
            .cloned()
            .filter(|heal| &heal.healed.uuid == player)
            .collect();

        damages_taken.sort_by_key(|d| d.date);
        damages_caused.sort_by_key(|d| d.date);
        heals.sort_by_key(|h| h.date);

        PlayerAlterationsAggregate {
            damages_taken_total: (&damages_taken)
                .iter()
                .fold(0u32, |acc, damage| damage.damage as u32 + acc),
            damages_caused_total: (&damages_caused)
                .iter()
                .fold(0u32, |acc, damage| damage.damage as u32 + acc),
            heals_total: (&heals)
                .iter()
                .fold(0u32, |acc, heal| heal.heal as u32 + acc),

            damages_taken,
            damages_caused,
            heals,

            kills: damages
                .iter()
                .filter(|damage| match &damage.cause {
                    DamageCause::Player(cause) => &cause.player.uuid == player && damage.lethal,
                    _ => false
                })
                .map(|damage| damage.damagee.clone())
                .collect(),

            killed_by: damages
                .iter()
                .filter(|damage| &damage.damagee.uuid == player && damage.lethal)
                .map(|damage| damage.cause.clone())
                .last(),

            game_duration: match damages
                .iter()
                .filter(|damage| &damage.damagee.uuid == player && damage.lethal)
                .map(|damage| &damage.date)
                .last()
            {
                Some(date) => since(date, begin),
                None => damages
                    .iter()
                    .last()
                    .map(|damage| since(&damage.date, begin))
                    .unwrap_or(Duration::from_secs(0)),
            },

            rank,
        }
    }

    fn aggregate_environmental_damages(damages: &Vec<Damage>) -> EnvironmentalDamagesAggregate {
        let mut aggregated = EnvironmentalDamagesAggregate {
            entities: BTreeMap::new(),
            environment: BTreeMap::new()
        };

        damages
            .iter()
            .filter(|damage| !matches!(damage.cause, DamageCause::Player { .. }))
            .for_each(|damage| {
                let aggregate = match damage.cause {
                    DamageCause::Entity(_) => &mut aggregated.entities,
                    _ => &mut aggregated.environment
                };

                let key = match &damage.cause {
                   DamageCause::Entity(cause) => cause.entity.clone(),
                    _ => damage.cause.to_string()
                };

                aggregate.insert(
                    key.clone(),
                    match aggregate.get(&key.clone()) {
                        Some(aggregate) => aggregate + damage.damage as u32,
                        None => damage.damage as u32,
                    },
                );
            });

        aggregated
    }
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct PlayerAlterationsAggregate {
    pub damages_taken: Vec<Damage>,
    pub damages_taken_total: u32,
    pub damages_caused: Vec<Damage>,
    pub damages_caused_total: u32,
    pub heals: Vec<Heal>,
    pub heals_total: u32,
    pub kills: Vec<SimplePlayer>,
    pub killed_by: Option<DamageCause>,
    pub game_duration: Duration,
    pub rank: u8,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct EnvironmentalDamagesAggregate {
    pub entities: BTreeMap<String, u32>,
    pub environment: BTreeMap<String, u32>
}
