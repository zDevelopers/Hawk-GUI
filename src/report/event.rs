use std::time::Duration;

use chrono::{DateTime, FixedOffset};
use uuid::Uuid;

use crate::report::raw::Event as RawEvent;
use crate::report::report::since;

#[inline(always)]
pub fn default_event_type() -> EventType {
    EventType::Blue
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Event {
    pub date: DateTime<FixedOffset>,
    pub since_beginning: Duration,

    #[serde(default = "default_event_type", rename = "type")]
    pub event_type: EventType,

    pub title: String,
    pub description: Option<String>,

    pub icon: EventIcon,
}

impl Event {
    pub fn from_raw(raw_event: RawEvent, begin: &DateTime<FixedOffset>) -> Self {
        Event {
            date: raw_event.date,
            since_beginning: since(&raw_event.date, begin),
            event_type: raw_event.event_type,
            title: raw_event.title,
            description: raw_event.description,
            icon: raw_event.icon,
        }
    }

    pub fn from_raw_vec(raw_events: Vec<RawEvent>, begin: &DateTime<FixedOffset>) -> Vec<Self> {
        let mut events: Vec<Self> = raw_events
            .into_iter()
            .map(|raw_event| Self::from_raw(raw_event, begin))
            .collect();

        events.sort_by_key(|e| e.date);

        events
    }
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum EventType {
    Blue,
    Gold,
    Green,
    Red,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "snake_case", tag = "type")]
pub enum EventIcon {
    Player { uuid: Uuid },
    Icon { icon_id: String },
    Url { url: String },
}
