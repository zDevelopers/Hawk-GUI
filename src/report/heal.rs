use std::collections::HashMap;
use std::rc::Rc;
use std::time::Duration;

use chrono::{DateTime, FixedOffset};
use uuid::Uuid;

use crate::report::errors::{InvalidReportError, ReportResult};
use crate::report::player::{Player, SimplePlayer};
use crate::report::raw::Heal as RawHeal;
use crate::report::report::since;

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Heal {
    pub date: DateTime<FixedOffset>,
    pub since_beginning: Duration,
    pub cause: HealCause,
    pub healed: SimplePlayer,
    pub heal: u16,
}

impl Heal {
    pub fn from_raw(
        raw_heal: RawHeal,
        players: &HashMap<Uuid, Rc<Player>>,
        begin: &DateTime<FixedOffset>,
    ) -> ReportResult<Self> {
        match players.get(&raw_heal.healed) {
            Some(healed) => Ok(Heal {
                date: raw_heal.date,
                since_beginning: since(&raw_heal.date, begin),
                cause: raw_heal.cause,
                healed: healed.into(),
                heal: raw_heal.heal,
            }),
            None => Err(InvalidReportError::MissingPlayerReference {
                uuid: raw_heal.healed,
            }),
        }
    }

    pub fn from_raw_vec(
        raw_heals: Vec<RawHeal>,
        players: &HashMap<Uuid, Rc<Player>>,
        begin: &DateTime<FixedOffset>,
    ) -> ReportResult<Vec<Self>> {
        let mut heals = raw_heals
            .into_iter()
            .map(|heal| Self::from_raw(heal, players, begin))
            .collect::<ReportResult<Vec<Self>>>()?;

        heals.sort_by_key(|h| h.date);

        Ok(heals)
    }
}

#[derive(Serialize, Deserialize, Debug, Clone)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum HealCause {
    Natural,
    GoldenApple,
    NotchApple,
    HealingPotion,
    Command,
    Unknown,
}
