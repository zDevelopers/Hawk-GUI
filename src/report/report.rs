use std::collections::HashMap;
use std::rc::Rc;
use std::result::Result::Ok;
use std::time::Duration;

use chrono::{DateTime, FixedOffset};
use uuid::Uuid;

use crate::report::*;
use crate::report::errors::ReportResult;

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Report {
    pub match_uuid: Uuid,
    pub title: String,
    pub date: DateTime<FixedOffset>,
    pub minecraft: Option<String>,
    pub settings: settings::Settings,
    pub players: Vec<player::Player>,
    pub teams: Vec<team::Team>,
    pub winners: Vec<player::SimplePlayer>,
    pub damages: Vec<damage::Damage>,
    pub heals: Vec<heal::Heal>,
    pub events: Vec<event::Event>,
    pub aggregates: aggregates::Aggregate,
    pub has_players_without_team: bool,
}

impl Report {
    ///
    /// Processes a raw report given by an user
    /// and returns a ready-to-be-serialized-and-used
    /// structure.
    ///
    pub fn from_raw(raw_report: raw::Report) -> ReportResult<Self> {
        let players_colors = {
            let mut players_colors = HashMap::new();

            for team in &raw_report.teams {
                for player in &team.players {
                    players_colors.insert(*player, team.color);
                }
            }

            players_colors
        };

        let settings = raw_report.settings;
        let teams = raw_report.teams;

        let players: HashMap<Uuid, Rc<player::Player>> = raw_report.players
            .into_iter()
            .map(|player| {
                (
                    player.uuid,
                    Rc::new(player::Player::from_raw(
                        player,
                        &teams,
                        &players_colors,
                        &settings.players,
                    )),
                )
            })
            .collect();

        let begin = raw_report.date;

        let mut damages = raw_report.damages;
        // Damage::from_raw_vec expect damages to be sorted by chronological order.
        damages.sort_by_key(|d| d.date);

        let damages = damage::Damage::from_raw_vec(damages, &players, &begin)?;
        let heals = heal::Heal::from_raw_vec(raw_report.heals, &players, &begin)?;

        let winners = match raw_report.winners {
            Some(winners) if !winners.is_empty() => {
                let mut team_players = winners.iter()
                    .map(|p| players.get(&p)
                        .map(Into::into)
                        .ok_or(errors::InvalidReportError::MissingPlayerReference { uuid: *p }))
                    .collect::<ReportResult<Vec<player::SimplePlayer>>>()?;

                team_players.sort_by(|a, b| a.name.cmp(&b.name));

                team_players
            },
            _ => Self::extract_winners(&players, &damages),
        };

        let aggregates = aggregates::Aggregate::from_raw(&players, &damages, &heals, &begin, &settings.players);

        let mut players_list: Vec<player::Player> = players
            .iter()
            .map(|(_, player)| player.as_ref().clone())
            .collect();

        players_list.sort_by(|a, b| a.name.cmp(&b.name));

        Ok(Report {
            match_uuid: raw_report.match_uuid,
            title: raw_report.title,
            date: raw_report.date,
            minecraft: raw_report.minecraft,
            settings,
            players: players_list,
            teams: team::Team::from_raw_vec(teams, &players)?,
            events: event::Event::from_raw_vec(raw_report.events, &begin),
            aggregates,
            winners,
            damages,
            heals,
            has_players_without_team: players
                .iter()
                .any(|(_uuid, player)| player.team.is_none()),
        })
    }

    fn extract_winners(
        players: &HashMap<Uuid, Rc<player::Player>>,
        damages: &Vec<damage::Damage>,
    ) -> Vec<player::SimplePlayer> {
        let deads: Vec<Uuid> = damages
            .iter()
            .filter(|damage| damage.lethal)
            .map(|damage| damage.damagee.uuid)
            .collect();

        let mut winners: Vec<player::SimplePlayer> = players
            .iter()
            .map(|(_uuid, player)| player)
            .filter(|player| !deads.contains(&player.as_ref().uuid))
            .map(|player| player.as_ref().into())
            .collect();

        winners.sort_by(|a, b| a.name.cmp(&b.name));

        winners
    }
}

pub fn since(now: &DateTime<FixedOffset>, before: &DateTime<FixedOffset>) -> Duration {
    (now.clone() - before.clone())
        .to_std()
        .unwrap_or(Duration::new(0, 0))
}
