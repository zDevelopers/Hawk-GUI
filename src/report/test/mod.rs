extern crate insta;

use std::fs;
use std::path::Path;

use insta::*;

use crate::*;
use crate::report::errors::ReportResult;
use crate::report::report::Report;

fn process_from_path(path: &Path) -> ReportResult<Report> {
    match fs::read_to_string(path) {
        Ok(input) => match serde_json::from_str(input.as_str()) {
            Ok(raw_report) => report::report::Report::from_raw(raw_report),
            Err(e) => panic!("Unable to parse raw report: {}", e)
        },
        Err(e) => panic!("Unable to read input report test file: {}", e)
    }
}

fn process(input: &'static str) -> ReportResult<Report> {
    process_from_path(Path::new(format!("./src/report/test/inputs/{}.json", input).as_str()))
}

fn process_to_string(input: &'static str) -> String {
    match process(input) {
        Ok(report) => match serde_json::to_string_pretty(&report) {
            Ok(json_report) => json_report,
            Err(e) => panic!("Unable to convert processed report to json: {}", e)
        }
        Err(e) => panic!("Unable to process report: {}", e)
    }
}

fn assert_input_with_snapshot(input: &'static str) {
    assert_snapshot!(input, process_to_string(input));
}

fn assert_input_fails(input: &'static str, message: &'static str) {
    assert!(process(input).is_err(), message);
}

#[test]
fn test_empty_report() {
    assert_input_with_snapshot("empty");
}

#[test]
fn test_minecraft_version() {
    assert_input_with_snapshot("minecraft_version");
}

#[test]
fn test_generator() {
    assert_input_with_snapshot("generator");
    assert_input_with_snapshot("generator_without_link");
}

#[test]
#[should_panic(expected = "missing field `name`")]
#[allow(unused_must_use)]
fn test_generator_without_name() {
    process("generator_without_name");
}

#[test]
fn test_broken_player_reference_should_fail() {
    assert_input_fails("broken_player_link", "Processing must fail if there is a broken player reference");
}
