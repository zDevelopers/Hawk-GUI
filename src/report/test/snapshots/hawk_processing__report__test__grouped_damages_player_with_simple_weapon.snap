---
source: src/report/test/mod.rs
expression: process_to_string(input)
---
{
  "match_uuid": "9be8ef14-a14e-4f96-b61e-b865c27ada8f",
  "title": "§5KTZ §dVII",
  "date": "2018-12-15T22:15:33+02:00",
  "minecraft": null,
  "settings": {
    "date": true,
    "players_count": true,
    "winners": true,
    "summary": {
      "enabled": true,
      "history": true,
      "players": true,
      "teams": true
    },
    "damages": {
      "enabled": true,
      "damages_per_players": true,
      "damages_per_team": true,
      "damages_from_environment": true,
      "display_killer": true
    },
    "players": {
      "enabled": true,
      "play_time": true,
      "global_statistics": true,
      "statistics_whitelist": [],
      "statistics_highlight": [],
      "used": false,
      "used_whitelist": [],
      "used_highlight": [],
      "mined": true,
      "mined_whitelist": [],
      "mined_highlight": [],
      "picked_up": true,
      "picked_up_whitelist": [],
      "picked_up_highlight": []
    },
    "generator": null
  },
  "players": [
    {
      "uuid": "55b89885-673c-4458-a8bd-33ad9c6190b4",
      "name": "Dada_exe",
      "color": "NONE",
      "team": null,
      "tag_line": "Allié",
      "tag_line_secondary": "",
      "tag_line_details": "Allié avec Jenjeur",
      "statistics": null,
      "displayed_statistics": null
    }
  ],
  "teams": [],
  "winners": [
    {
      "uuid": "55b89885-673c-4458-a8bd-33ad9c6190b4",
      "name": "Dada_exe",
      "color": "NONE",
      "team": null
    }
  ],
  "damages": [
    {
      "date": "2018-12-15T22:59:16+02:00",
      "since_beginning": {
        "secs": 2623,
        "nanos": 0
      },
      "cause": {
        "type": "player",
        "player": {
          "uuid": "55b89885-673c-4458-a8bd-33ad9c6190b4",
          "name": "Dada_exe",
          "color": "NONE",
          "team": null
        },
        "weapon": {
          "id": "minecraft:netherite_sword",
          "Count": 1,
          "tag": null
        }
      },
      "damagee": {
        "uuid": "55b89885-673c-4458-a8bd-33ad9c6190b4",
        "name": "Dada_exe",
        "color": "NONE",
        "team": null
      },
      "damage": 12,
      "lethal": false
    }
  ],
  "heals": [],
  "events": [],
  "aggregates": {
    "global_statistics": {
      "generic": {
        "visible": [],
        "hidden": []
      },
      "used": null,
      "mined": {
        "visible": [],
        "hidden": []
      },
      "picked_up": {
        "visible": [],
        "hidden": []
      }
    },
    "players_damages": {
      "55b89885-673c-4458-a8bd-33ad9c6190b4": {
        "damages_taken": [
          {
            "date": "2018-12-15T22:59:16+02:00",
            "since_beginning": {
              "secs": 2623,
              "nanos": 0
            },
            "cause": {
              "type": "player",
              "player": {
                "uuid": "55b89885-673c-4458-a8bd-33ad9c6190b4",
                "name": "Dada_exe",
                "color": "NONE",
                "team": null
              },
              "weapon": {
                "id": "minecraft:netherite_sword",
                "Count": 1,
                "tag": null
              }
            },
            "damagee": {
              "uuid": "55b89885-673c-4458-a8bd-33ad9c6190b4",
              "name": "Dada_exe",
              "color": "NONE",
              "team": null
            },
            "damage": 12,
            "lethal": false
          }
        ],
        "damages_taken_total": 12,
        "damages_caused": [
          {
            "date": "2018-12-15T22:59:16+02:00",
            "since_beginning": {
              "secs": 2623,
              "nanos": 0
            },
            "cause": {
              "type": "player",
              "player": {
                "uuid": "55b89885-673c-4458-a8bd-33ad9c6190b4",
                "name": "Dada_exe",
                "color": "NONE",
                "team": null
              },
              "weapon": {
                "id": "minecraft:netherite_sword",
                "Count": 1,
                "tag": null
              }
            },
            "damagee": {
              "uuid": "55b89885-673c-4458-a8bd-33ad9c6190b4",
              "name": "Dada_exe",
              "color": "NONE",
              "team": null
            },
            "damage": 12,
            "lethal": false
          }
        ],
        "damages_caused_total": 12,
        "heals": [],
        "heals_total": 0,
        "kills": [],
        "killed_by": null,
        "game_duration": {
          "secs": 2623,
          "nanos": 0
        },
        "rank": 1
      }
    },
    "environmental_damages": {
      "entities": {},
      "environment": {}
    }
  },
  "has_players_without_team": true
}
